sound/soc/intel/boards/sof_sdw.c

```diff
@@ -515,6 +515,43 @@ int sdw_trigger(struct snd_pcm_substream *substream, int cmd)
 	return ret;
 }
 
+int sdw_hw_params(struct snd_pcm_substream *substream,
+		  struct snd_pcm_hw_params *params)
+{
+	struct snd_soc_pcm_runtime *rtd = asoc_substream_to_rtd(substream);
+	struct snd_soc_dai *cpu_dai;
+	struct snd_soc_dai *codec_dai;
+	unsigned int ch_mask;
+	int step;
+	int i;
+	int j;
+
+	if (!rtd->dai_link->codec_ch_maps)
+		return 0;
+
+	if (substream->stream == SNDRV_PCM_STREAM_PLAYBACK) {
+		step = 0;
+		ch_mask = GENMASK(params_channels(params) - 1, 0);
+	} else {
+		step = rtd->dai_link->num_codecs;
+		ch_mask = GENMASK(params_channels(params) / step - 1, 0);
+	}
+
+	/*
+	 * The captured data will be combined from each cpu DAI if the dai
+	 * link has more than one codec DAIs. Set codec channel mask and
+	 * ASoC will set the corresponding channel numbers for each cpu dai.
+	 */
+	for_each_rtd_cpu_dais(rtd, i, cpu_dai) {
+		for_each_rtd_codec_dais(rtd, j, codec_dai) {
+			if (rtd->dai_link->codec_ch_maps[j].connected_cpu_id != i)
+				continue;
+			rtd->dai_link->codec_ch_maps[j].ch_mask = ch_mask << (j * step);
```

Comment:
I think this is only working in the corner case of `step = 0` and if
`(params_channels(params) / rtd->dai_link->num_codecs) == rtd->dai_link->num_codecs`
if the stream is 8 channel and you have 2 codecs (each capable of 4 channel) then you will have overlap on bit 2-3 of the masks?
Similarly, if you have two mono codecs and open a stereo capture, then there will be a gap and bit 1 will be not claimed for any codec?

Reply:
> I think this is only working in the corner case of `step = 0` and if `(params_channels(params) / rtd->dai_link->num_codecs) == rtd->dai_link->num_codecs` if the stream is 8 channel and you have 2 codecs (each capable of 4 channel) then you will have overlap on bit 2-3 of the masks? Similarly, if you have two mono codecs and open a stereo capture, then there will be a gap and bit 1 will be not claimed for any codec?

@ujfalusi Thanks for pointing this out. Much appreciated. The step should be `params_channels(params) / num_codecs` instead. It makes me think about what if channels < num_codecs? @plbossart Should we handle the channels < num_codecs case?

changes: [0503.txt](https://github.com/thesofproject/linux/files/11381118/0503.txt)

Reply:
> Not fully following the question @bardliao.
> 
> Are you asking what would happen if the application wanted a mono channel but there are two codecs capable of providing captured data?

Yes, for example, there are 4 amps, and the application asks for 2 channels capture.

> 
> I think that's a configuration bug, there would be no way to select data from one codec or the other. This sort of selection would need to happen with a mux or channel selector with an explicit configuration on what to select

Do we need to return error and/or show error log in machine driver?

Reply:
I am not sure what would happen. This would be a hw_params error, no? I don't think we need to worry too much about this case.

Reply:
Does the position of the masks matter or only the number of bits? In core it looks that way, iow:
b00001111 is treated identically to b00110011.
Is it so?

IOW, core does not care and know anything about the source of each channel?
How do we know  that we are interleaving the audio channels (codec1-ch1, codec2-ch1, codec1-ch2, codec2-ch2) or not interleaving (codec1-ch1, codec1-ch2, codec2-ch1, codec2-ch2)?

Reply:
> Does the position of the masks matter or only the number of bits? In core it looks that way, iow: b00001111 is treated identically to b00110011. Is it so?

Only the number of bits matter.

> 
> IOW, core does not care and know anything about the source of each channel? How do we know that we are interleaving the audio channels (codec1-ch1, codec2-ch1, codec1-ch2, codec2-ch2) or not interleaving (codec1-ch1, codec1-ch2, codec2-ch1, codec2-ch2)?

For SOF IPC4, it is set in [sof_ipc4_prepare_copier_module](https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/sof/ipc4-topology.c#L1637)

Reply:
hrm, so, how does this differs from the tdm_mask?
codec1's capture_tdm_mask:   b00000011
codec1's playback_tdm_mask: b00000011
codec2's capture_tdm_mask:   b00001100
codec2's playback_tdm_mask: b00000011

cpu capture mask is codec1|codec2 = b00001111 -> four channels
cpu playback mask is codec1|codec2 = b00000011 -> stereo channels

if you want to play different channels to the codecs, then:
codec1's playback_tdm_mask: b00000011
codec2's playback_tdm_mask: b00001100
cpu playback mask is codec1|codec2 = b00001111 -> four channels

I really don't see any difference apart that with the `codec_ch_maps[x].ch_mask` we completely ignore the mask nature of the ch_mask.

I got the argument that the tdm_mask predates SDW but I don't think it is that different as a concept. If you have multiple codecs on the SDW or I2S bus you need to configure them in a way that they will not conflict on the bus, right?

If we overlook the corner case of multiple CPUs connected to single codec, which is technically possible with I2S, we can make an assumption that a codec is only connected to one CPU (for now).

Instead of `rtd->dai_link->codec_ch_maps[]` we can have `rtd->dai_link->codec_to_cpu_map[]` as an integer array, set up the tdm_mask (playback and capture) of the codecs and based on them set up the tdm_mask for the CPU side (or together the connected codec's masks).
This will give the number of channels and it will preserve the channel placement on the bus if it is needed by the underlying physical bus, like I2S.

For the capture case we are aiming here, where we always want to receive from both codecs:
four channel:
codec1's capture_tdm_mask:   b00000011
codec2's capture_tdm_mask:   b00001100
cpu capture mask is codec1|codec2 = b00001111 -> four channels

Stereo
codec1's capture_tdm_mask:   b00000001
codec2's capture_tdm_mask:   b00000010
cpu capture mask is codec1|codec2 = b00000011

But for Stereo one can say to only capture from codec1.:
codec1's capture_tdm_mask:   b00000011
codec2's capture_tdm_mask:   b00000000
cpu capture mask is codec1|codec2 = b00000011

@plbossart, @bardliao, what do you think?

Reply:
The problem with TDM mask is that there is a notion of TX_MASK and RX_MASK, which is built-in the API. 

````c
int (*set_tdm_slot)(struct snd_soc_dai *dai,
		unsigned int tx_mask, unsigned int rx_mask,
		int slots, int slot_width);
````

We don't want to tie capture and playback, and we don't want a nothing of slot_width either. Capture and playback are completely independent in SoundWire.

So yes, channel_mask and tdm_mask are both masks, but that's about the end of the commonality.

TDM is a very specific concept, it would be very confusing IMHO to understand.

Reply:
And in addition, the TDM concept only works with one CPU DAI and N Codec DAIs. There's no way for TDM to support multiple CPU DAIs, it's not supported at all

Reply:
> And in addition, the TDM concept only works with one CPU DAI and N Codec DAIs. There's no way for TDM to support multiple CPU DAIs, it's not supported at all

I did had a setup with minimal not upstreamed patches to support it with I2S.
cpu1 capture TDM mask: b00000011
cpu2 capture TDM mask: b00001100
codec capture TDM maks: b00001111

PCM1 (cpu1) would record only slot 0 and 1
PCM2 (cpu2) would record only slot 2 and 3
in both cases the codec sends four channel audio on slot 0 - 3.
Th tricky part is to graph up the clock generator in such a setup (in this simple case it was the codec).

Reply:
> The problem with TDM mask is that there is a notion of TX_MASK and RX_MASK, which is built-in the API.
> 
> ```c
> int (*set_tdm_slot)(struct snd_soc_dai *dai,
> 		unsigned int tx_mask, unsigned int rx_mask,
> 		int slots, int slot_width);
> ```
> 
> We don't want to tie capture and playback, and we don't want a notion of slot_width either. Capture and playback are completely independent in SoundWire.

I agree that the API ties tx and rx masks which is not ideal for I2S either as we have playback/capture only 'codecs', the TX does not make much sense on a capture only codec, right?
And to be fair, I2S also have separate data line for TX and RX, only the FS and BCLK is shared, so they are kind of independent as well, no?
 
> So yes, channel_mask and tdm_mask are both masks, but that's about the end of the commonality.

and channel_mask as it is introduced is not a mask, it is the number of bits in an `unsigned int` and it is very well SDW specific thing. It is just telling  that if the capture is 4 channel, take 2 from one codec and take 2 from the other, but it is not saying _which_ channel and where those 2 channels from the codecs are going to be ending up in the 4 channel recording, right?
 
> TDM is a very specific concept, it would be very confusing IMHO to understand.

Applying TDM to non DSP_A/B formats is confusing on the I2S bus, but I don't see that with the SDW. It is already abstracted in a way that  the underlying bus is not visible in machine level, we have CODECs / CPUs supporting x number of channels and they will match up under the hood, in this level we don't know/care where those channels are transmitted on the bus.

I find it equally confusing to introduce a new, competing, but not connected set of information for the same thing.
How non SDW setup supposed to use this? I don't see that...

Reply:
> 
> I agree that the API ties tx and rx masks which is not ideal for I2S either as we have playback/capture only 'codecs', the TX does not make much sense on a capture only codec, right? And to be fair, I2S also have separate data line for TX and RX, only the FS and BCLK is shared, so they are kind of independent as well, no?

some implementations have different bclock and fs for playback and capture so that's even less clear... I2S/TDM is the standard that keeps on giving with an infinite degrees of freedom

> > So yes, channel_mask and tdm_mask are both masks, but that's about the end of the commonality.
> 
> and channel_mask as it is introduced is not a mask, it is the number of bits in an `unsigned int` and it is very well SDW specific thing. It is just telling that if the capture is 4 channel, take 2 from one codec and take 2 from the other, but it is not saying _which_ channel and where those 2 channels from the codecs are going to be ending up in the 4 channel recording, right?

The combination is clearly a bit-wise operation, see the code excerpt here

````c
		/*
		 * construct cpu channel mask by combining ch_mask of each
		 * codec which maps to the cpu.
		 */
		for_each_rtd_codec_dais(rtd, j, codec_dai) {
			if (rtd->dai_link->codec_ch_maps[j].connected_cpu_id == i)
				ch_mask |= rtd->dai_link->codec_ch_maps[j].ch_mask;
		}
````

And you absolutely control how the individual codecs contribute to the 4 channel recording. 

If codec0 uses mask 0xC and codec1 uses 0x3, then the channels from codec 1 will be included in channels 1..0.

What the mask operation does not handle is reorder the channels provided by each codec. That should probably be clearer in the commit message.

> 
> > TDM is a very specific concept, it would be very confusing IMHO to understand.
> 
> Applying TDM to non DSP_A/B formats is confusing on the I2S bus, but I don't see that with the SDW. It is already abstracted in a way that the underlying bus is not visible in machine level, we have CODECs / CPUs supporting x number of channels and they will match up under the hood, in this level we don't know/care where those channels are transmitted on the bus.
> 
> I find it equally confusing to introduce a new, competing, but not connected set of information for the same thing. How non SDW setup supposed to use this? I don't see that...

It's a very odd logic you're using @ujfalusi ...

It's not competing and it's not the same thing. TDM rx/tx masks are not even suitable for all TDM implementations where the number of slots is different between RX and TX.
 
SoundWire is the only case with multiple cpu-dais today, I don't see how we could reject an extension made specifically for multiple cpu-dais on the grounds that only SoundWire would use it. 

I view it the other way around: this patch removes an old framework limitation with M:N mappings between cpu and codec dais. I just happens to be very useful for SoundWire. It was not designed specifically for SoundWire, there is nothing in the patch that @bardliao suggested that is SoundWire-specific.

Reply:
BTW even for the TDM case, the only thing that the RX and TX masks tell is which slot is used by what codec. If a codec uses multiple slots, there's nothing that really tells which channel maps to what slot. This is not something that can be controlled at the link level, usually there's a separate control for this.

The other point is that in the SoundWire case, we *could* interleave channels generated by different codecs, but that would require the use of a specific data port mode where channels are spaced out in the frame instead of being transmitted in a contiguous manner. 

What is suggested by @bardliao would work, e.g. if the codec0 use mask 0x5 and codec1 used mask 0xa we could interleave the codec0 and codec1 channels.
That's not supported in the Intel bit allocation though, so that's only a theoretical possibility.

Reply:
OK.

Reply:
So, OK with the following note: in non SDW case the machine driver will set up this mapping based on the TDM masks or have a helper in core to do so. If the need arises, that is.